kep/podgroup-coscheduling/README.md

```diff
@@ -0,0 +1,140 @@
+# Coscheduling based on PodGroup CRD
+
+## Table of Contents
+
+<!-- toc -->
+- [Coscheduling based on PodGroup CRD](#Coscheduling based on PodGroup CRD)
+  - [Table of Contents](#table-of-contents)
+  - [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+  - [Design Details](#design-details)
+    - [PodGroup](#podgroup)
+    - [Controller](#controller)
+    - [Extension points](#extension-points)
+      - [QueueSort](#queuesort)
+      - [PreFilter](#prefilter)
+      - [Permit](#permit)
+      - [PostBind](#postbind)
+  - [Known Limitations](#Known Limitations)
+<!-- /toc -->
+
+## Motivation
+Currently, through the default scheduler of Kubernetes, we cannot ensure a group of pods scheduled at the same time . Under some scene, it would waste resources since some pods need work together, like spark, tensorflow and so on . So, podgroup-coscheduling is aimed at solving the issue.
+
+## Goals
+1. Base on scheduling framework, implementing the gang scheduling.
+2. Define a CRD name PodGroup to help pod scheduling.
+ 
+## Non-Goals
+Sort the job when we submit jobs to a cluster. Currently, we can only do this base on pods.
+ 
+## Use Cases
+Spark jobs, tensorflow jobs and other pods have to run together.
+
+## Design Details
+
+### PodGroup
+
+We define a CRD name PodGroup to help schedule, its definition is as follows:
+```build
+// PodGroupSpec represents the template of a pod group.
+type PodGroupSpec struct {
+	// MinMember defines the minimal number of members/tasks to run the pod group;
+	// if there's not enough resources to start all tasks, the scheduler
+	// will not start anyone.
+	MinMember uint32 `json:"minMember"`
+
+	// If specified, indicates the PodGroup's priority. "system-node-critical" and
+	// "system-cluster-critical" are two special keywords which indicate the
+	// highest priorities with the former being the highest priority. Any other
+	// name must be defined by creating a PriorityClass object with that name.
+	// If not specified, the PodGroup priority will be default or zero if there is no
+	// default.
+	// +optional
+	PriorityClassName string `json:"priorityClassName,omitempty"`
+
+	// MinResources defines the minimal resource of members/tasks to run the pod group;
+	// if there's not enough resources to start all tasks, the scheduler
+	// will not start anyone.
+	MinResources *v1.ResourceList `json:"minResources,omitempty"`
+
+	// MaxScheduleTime defines the maximal time of members/tasks to wait before run the pod group;
+	MaxScheduleTime *metav1.Duration `json:"maxScheduleTime,omitempty"`
```

Comment:
We should avoid defining API as Go type `Duration`, use `int64` instead, and rename it like `ScheduleTimeoutSeconds` or `WaitTimeoutSeconds`.

Ref:

- [api-conventions.md#units](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#units)
- [api-conventions.md#naming-conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#naming-conventions)